File: back-end/src/modules/product/mappers/product-detail.mapper.ts
/**
 * @description Product detail mapper
 * @author Nhut Tan
 * @since 2025-09-24
 * @version 1.0.0
 */
import { Injectable } from '@nestjs/common';
import { ProductDetailsEntity } from '../entities/product-details.entity';
import { ProductDetailResponseDto } from '../dtos/product-detail-response.dto';
import { ProductImageEntity } from '../../image/entities/product-image.entity';

@Injectable()
export class ProductDetailMapper {
  toProductDetailResponse(
    productDetailEntity: ProductDetailsEntity,
  ): ProductDetailResponseDto {
    return {
      id: productDetailEntity.id,
      name: productDetailEntity.product.name,
      price: productDetailEntity.product.price,
      discount: productDetailEntity.product.discount,
      rating: productDetailEntity.rating,
      size: productDetailEntity.size.split(', '),
      color: productDetailEntity.color,
      description: productDetailEntity.description,
      imageList: productDetailEntity.product.productImages.map(
        (productImageEntity: ProductImageEntity): string => {
          return productImageEntity.image.url;
        },
      ),
    };
  }
}

File: ECommerce/app/src/main/java/com/example/e_commerce/Mappers/ProductDetailMapper.kt
/**
 * @description Product mapper object
 * @author @nhuttan12
 * @since 2025-24-12
 * @modifies 2026-01-07
 * @version 1.0.1
 */

package com.example.e_commerce.Mappers

import com.example.e_commerce.DTOs.ProductDetailDTO
import com.example.e_commerce.Helper.StringConverter
import com.example.e_commerce.Model.ProductDetailModel

object ProductDetailMapper {
    private val stringConverter: StringConverter = StringConverter()
    fun fromDto(dto: ProductDetailDTO): ProductDetailModel {
        return ProductDetailModel(
            title = dto.name,
            description = this.stringConverter.splitDescription(dto.description),
            picUrl = ArrayList(dto.imageList),
            size = ArrayList(dto.size),
            color = ArrayList(dto.color),
            price = applyDiscount(dto.price.toDouble(), dto.discount),
            oldPrice = dto.price.toDouble(),
            rating = dto.rating,
            numberInCart = 1
        )
    }
//
//    fun fromDtoList(dtos: List<ProductDetailDTO>): ArrayList<ProductDetailModel> {
//        return dtos.map { fromDto(it) }.toCollection(ArrayList())
//    }

    fun applyDiscount(price: Double, discount: Int): Double {
        return if (discount in 1..100) {
            price * (100 - discount) / 100
        } else {
            price
        }
    }
}